### PR TITLE
[search] Search result streaming

### PR DESCRIPTION
**Pull request by ejgallego**
The current `Search.search_*` API returns a large `Pp.std_ppcmds`
object with the formatting of the search results. For large searches this is inefficient as:
- a large list _pretty printing_ object is generated only to be consumed later.
- users must wait for the whole search to complete before seeing any results. 

This PR modifies the API so users can perform per-result processing of
search results. We now emit a feedback message for every result, thus streaming follows.

This PR offers nice improvements to JsCoq, now search
results are printed as they are generated.

This is a new version of #147 due to change of target branch.

### CODE REVIEW

**Comment by maximedenes** [comment]
Argh! Just when I was about to branch. And of course, you want that in 8.6, right? :stuck_out_tongue:

**Comment by ejgallego** [comment]
Yeah sorry for that, I didn't manage to properly test the old PR #147 until today, as I didn't have good client side support for it.

IMO I don't care, it is a very minor improvement for "regular" search users. Ssreflect's search streams results fine without this.

**Comment by ejgallego** [comment]
Note the rebase to fix a wrong comment. Given how localized the change i I don't expect a lot of problem honestly.

**Comment by aspiwack** [comment]
Is there anything blocking this PR? @gmalecha any further/closing thoughts?

**Review by mattam82 — APPROVED** [review]
LGTM. From what I understand, the print name only could be done outside of Coq now thanks to the parameterization of the display function, is that right?

**Comment by ejgallego** [comment]
Rebased. @mattam82 , it could be also done before but with more effort as you had to use the `generic_search` interface. This is what ssreflect did. IMHO it makes sense to have that in the toplevel itself as it is a pretty specific option. Other toplevels don't use this method to query for completions, due to efficiency.

**Comment by ejgallego** [comment]
I am testing the rebase so please hold a second.

**Comment by ejgallego** [comment]
Tested. I got alarmed by a few test suite failures but they seem to be due to the notations warnings stuff.

File: toplevel/vernacentries.ml [hunk 1]
```diff
@@ -1665,6 +1665,24 @@ let interp_search_about_item env =
 	errorlabstrm "interp_search_about_item"
           (str "Unable to interp \"" ++ str s ++ str "\" either as a reference or as an identifier component")
 
+(* EG: I have to disagree with
```
Comment (maximedenes) [inline toplevel/vernacentries.ml]: Should I leave this comment here or remove it when merging?

Comment (ejgallego) [inline toplevel/vernacentries.ml]: Oh indeed let me rephrase it

**Comment by ejgallego** [comment]
Done, thanks for the comment.

**Comment by ejgallego** [comment]
Thanks!
